replica/table.cc

```diff
@@ -2805,7 +2805,53 @@ class compaction_group::table_state : public compaction::table_state {
         return _t.is_auto_compaction_disabled_by_user();
     }
     bool tombstone_gc_enabled() const noexcept override {
-        return _t._tombstone_gc_enabled;
+        /**
+         * #14870 - block our GC iff there is potential data
+         * left in commitlog that could cause resurrections
+         * on crash+replay.
+         * 
+         * We take a pessimistic approach and check for lowest
+         * rp still left in CL, and if any sstable we process 
+         * has high RP higher or equal than this, we must 
+         * block GC for now.
+         * 
+         * Note: this messes with several tests for compaction
+         * that expect a flush + compaction + check for data to
+         * have done GC already. See database.cc for workaround
+         * for this. 
+        */
+        if (!_t._tombstone_gc_enabled) {
+            return false;
+        }
+        auto cl = _t.commitlog();
+        if (!cl) {
+            return true;
+        }
+        auto lowrp = cl->lowest_active_table_rp(schema()->id());
+        if (!lowrp) {
+            return true;
+        }
+        bool ok = true;
+        auto check = [&](const shared_sstable& s) {
+            if (*lowrp <= s->get_stats_metadata().position) {
+                if (tlogger.is_enabled(log_level::debug)) {
+                    auto active = cl->active_cf_ids();
+                    tlogger.debug("GC blocked for {}:{} ({}). SSTable rp {} <= CL low mark {}",
+                        schema()->ks_name(), schema()->cf_name(), schema()->id(),
+                        s->get_stats_metadata().position, *lowrp
+                    );
+                    tlogger.debug("Active IDs (blocking): {}", active);
+                }
+
+                ok = false;
+                return stop_iteration::yes;
```

Comment:
This looks expensive, and my intuition says it will usually return false (for many tables/tablets), no proof yet.

Reply:
The lookup is not very expensive, though obviously o(nsstables + nsegments). As for ratio between true/false, not sure. If compaction only selects older sstables for GC:ing compaction, it should be pretty safe (for real world cases), but obviously (as stated elsewhere in this PR), for tests which create artificial compactions early etc we end up with lock-outs quite a lot. 

Which illustrates the annoying balance here I guess... (Note, the truncation-using solution does not suffer from this issue, but generates extra writes and flushes - even if those could potentially be alleviated at least partially).

Reply:
Now I understand the fix better. 

We call tombstone_gc_enabled() for every partition containing an expired tombstone.

So the complexity is

O(partitions_with_tombstone * (nsstables + nsegments))

I am worried about the overhead for tombstone heavy workloads (can be default TTL, not only regular deletion), which is common. And those with small partitions will suffer even more.

we can probably mitigate it by doing commitlog check only once per compaction. worse granularity, but less overhead.

@elcallio why don't we use replay position stored in sstables to avoid replaying older data?

Reply:
That is a good question. While the sstable only stores high RP (taken from memtable flushed), technically we could maybe use it the same way as her, but in inverse: I.e. if a higher RP that being replayed is in any live sstable, we can ignore the replay. Might be problematic however if a sstable is damaged/removed and we somehow rely on a combo of replay + repair to fix things up... Maybe that is never an issue...
I am probably forgetting something here...

Reply:
I believe the replay positions are useless. They tell us what commitlog data this sstable contains, but it doesn't allows us to exclude all commits before that position, because sstables contain filtered commitlog data. It's filtered by table (and that's something we know because we have the sstable->table association), but it can be filtered by token range (for compaction groups), but timestamp (for TWCS), and probably for other things. So to actually filter commits during replay, we need to know all the conditions that were used to create the sstable.

Reply:
Actually, as long as sstables always retain the highest RP contained, it should be fine, since commit log is guarding against loosing data not written to sstable, not data potentially lost due to sstable being lost (a latter matter). A higher RP in any sstable always means that all lower RP:s have been written to disk (remember, we do not allow reordering when writing to memtable) at some point, in some sstable, thus data in CL should be safe to skip one way or another. 

In the truncation case, obviously this data is useless since sstables are gone, hence the truncation table. I still have a nagging memory that there is some reason we don't use sstable positions, that theý are not quite reliable, but I can't remember if/why... There is also the issue again of resharding, which makes them harder to compare...